### PR TITLE
Support s3 api copy-object across repositories

### LINKED ISSUE #2162: Support for s3 api copy-object across buckets
Currently s3 api doesn't support copying objects across buckets

I am trying to copy  objects across buckets by running this using local storage  
`aws --endpoint-url http://s3.local.lakefs.io:8000 s3api copy-object --bucket demo-2 --key main/mytest7.py --copy-source demo-1/main/mytest.py`
and returns below error
`An error occurred (InvalidArgument) when calling the CopyObject operation: Copy Source must mention the source bucket and key: sourcebucket/sourcekey. ` 
I am following https://docs.aws.amazon.com/cli/latest/reference/s3api/copy-object.html#examples

### PR DESCRIPTION
**Pull request by eden-ohana**
Fixes #2162 

Currently supports copy-object only for the same repository.

### CODE REVIEW

File: pkg/gateway/operations/putobject.go [hunk 1]
```diff
@@ -32,75 +32,132 @@ const (
...
+	if err != nil {
+		o.Log(req).WithError(err).Error("could not read copy source")
+		_ = o.EncodeError(w, req, gatewayErrors.Codes.ToAPIErr(gatewayErrors.ErrInvalidCopySource))
+		return nil
+	}
+	return ent
+}
+
+// extractAndCopyEntryFromCopyReq: create copy of the file
+func extractAndCopyEntryFromCopyReq(w http.ResponseWriter, req *http.Request, o *PathOperation, copySource string) *catalog.DBEntry {
```
Comment (guy-har) [inline pkg/gateway/operations/putobject.go:85]: Doesn't really extract, this is basically a copy command that returning, please change name and documentation to reflect that

File: pkg/gateway/operations/putobject.go [hunk 2]
```diff
@@ -32,75 +32,132 @@ const (
...
+					Resource: permissions.ObjectArn(repoID, destPath)},
+			},
+			{
+				Permission: permissions.Permission{
+					Action:   permissions.ReadObjectAction,
+					Resource: permissions.ObjectArn(p.Repo, p.Path)},
+			}}}, nil
 }
 
+// extractEntryFromCopyReq: update metadata to refer to the source hash in the destination workspace
```
Comment (guy-har) [inline pkg/gateway/operations/putobject.go:67]: What is updated?

**Review by guy-har — CHANGES_REQUESTED** [review]
Grew out to be a very valuable PR,
Great Job!
Some minor requests. requesting changes mainly because of the panic returned in the default case.

**Comment by guy-har** [comment]
Not part of this PR, but maybe we should open a new Issue for adding copy to our API, we already have the capability. WDYT?

File: pkg/gateway/operations/putobject.go (hunk 2 above)
Comment (eden-ohana) [reply pkg/gateway/operations/putobject.go:66]: I divided it into two functions

File: nessie/s3_gateway_test.go [hunk 3]
```diff
@@ -104,3 +106,133 @@ func TestS3UploadAndDownload(t *testing.T) {
...
+	download, err := minioClient.GetObject(
+		ctx, repo, o.DestPath, minio.GetObjectOptions{})
+	if err != nil {
+		t.Errorf("minio.Client.GetObject(%s): %s", o.DestPath, err)
+	}
+	// compere files content
+	contents := bytes.NewBuffer(nil)
+	_, err = io.Copy(contents, download)
+	if err != nil {
+		t.Errorf("download %s: %s", o.DestPath, err)
```
Comment (guy-har) [inline nessie/s3_gateway_test.go:167]:
```suggestion
		t.Fatal("download %s: %s", o.DestPath, err)
```
You could use Fatal in cases of errors where you don't want your test to continue running

**Review by guy-har — APPROVED** [review]
Approved :)
Please remove or change the documentation I commented about.

File: nessie/s3_gateway_test.go [hunk 4]
```diff
@@ -104,3 +106,133 @@ func TestS3UploadAndDownload(t *testing.T) {
...
+	if strings.Compare(contents.String(), o.Content) != 0 {
+		t.Errorf(
+			"Downloaded bytes %v from uploaded bytes %v", contents.Bytes(), o.Content)
+	}
+
+	resp, err := client.StatObjectWithResponse(ctx, repo, mainBranch, &api.StatObjectParams{Path: "data/source-file"})
+	if err != nil {
+		t.Errorf("client.StatObject(%s): %s", o.SourcePath, err)
+	}
+	sourceObjectStats := resp.JSON200
```
Comment (guy-har) [inline nessie/s3_gateway_test.go:178]:
```suggestion
	require.Equal(t, http.StatusOK, resp.StatusCode())
	sourceObjectStats := resp.JSON200
```

File: nessie/s3_gateway_test.go [hunk 5]
```diff
@@ -104,3 +106,133 @@ func TestS3UploadAndDownload(t *testing.T) {
...
+	})
+	if err != nil {
+		t.Fatalf("minio.New: %s", err)
+	}
+
+	o := Object{
+		Content:    testutil.RandomString(rand, randomDataContentLength),
+		SourcePath: prefix + "source-file",
+		DestPath:   prefix + "dest-file",
+	}
```
Comment (guy-har) [inline nessie/s3_gateway_test.go:138]:
```suggestion

	Content:=    testutil.RandomString(rand, randomDataContentLength)
	SourcePath:= prefix + "source-file"
	DestPath: =  prefix + "dest-file"

```

Comment (guy-har) [reply nessie/s3_gateway_test.go:138]: You don't really need the Object type, also it's a bit confusing that the Object has two paths

**Pull request merged**